adapters/repos/db/lsmkv/roaringset/serialization.go

```diff
@@ -0,0 +1,194 @@
+// contains a large amount of data.
+//
+// The internal structure of the data is:
+//
+//	byte begin-start    | description
+//	--------------------|-----------------------------------------------------
+//	0-8                 | uint64 indicating the total length of the node,
+//	                    | this is used in cursors to identify the next node.
+//	8-16                | uint64 length indicator for additions sraor bm -> x
+//	16-(x+16)           | additions bitmap
+//	(x+16)-(x+24)       | uint64 length indicator for deletions sroar bm -> y
+//	(x+24)-(x+y+24)     | deletions bitmap
+//	(x+y+24)-(x+y+28)   | uint32 length indicator for primary key length -> z
+//	(x+y+28)-(x+y+z+28) | primary key
+type SegmentNode struct {
+	// Offset is not persisted in the node itself, but it is respected when
+	// building a [segmentindex.Key] using [*SegmentNode.KeyIndexAndWriteTo].
+	// multiple nodes
+	Offset int
+
+	data []byte
+}
+
+// Len indicates the total length of the [SegmentNode]. When reading multiple
+// segments back-2-back, such as in a cursor situation, the offset of element
+// (n+1) is the offset of element n + Len()
+func (sn *SegmentNode) Len() uint64 {
+	return binary.LittleEndian.Uint64(sn.data[0:8])
+}
+
+// Additions returns the additions roaring bitmap with shared state. Only use
+// this method if you can garantuee that you will only use it while holding a
+// maintenance lock or can otherwise be sure that no compaction can occur. If
+// you can't guarantee that, instead use [*SegmentNode.AdditionsWithCopy].
+func (sn *SegmentNode) Additions() *sroar.Bitmap {
+	additionsLength := binary.LittleEndian.Uint64(sn.data[8:16])
+	return sroar.FromBuffer(sn.data[16 : 16+additionsLength])
+}
+
+// AdditionsWithCopy returns the additions roaring bitmap without sharing state. It
+// creates a copy of the underlying buffer. This is safe to use indefinitely,
+// but much slower than [*SegmentNode.Additions] as it requires copying all the
+// memory. If you know that you will only need the contents of the node for a
+// duration of time where a lock is held that prevents compactions, it is more
+// efficient to use [*SegmentNode.Additions].
+func (sn *SegmentNode) AdditionsWithCopy() *sroar.Bitmap {
+	additionsLength := binary.LittleEndian.Uint64(sn.data[8:16])
+	return sroar.FromBufferWithCopy(sn.data[16 : 16+additionsLength])
+}
+
+// Deletions returns the deletions roaring bitmap with shared state. Only use
+// this method if you can garantuee that you will only use it while holding a
+// maintenance lock or can otherwise be sure that no compaction can occur. If
+// you can't guarantee that, instead use [*SegmentNode.DeletionsWithCopy].
+func (sn *SegmentNode) Deletions() *sroar.Bitmap {
+	additionsLength := binary.LittleEndian.Uint64(sn.data[8:16])
+	offset := additionsLength + 16
+	deletionsLength := binary.LittleEndian.Uint64(sn.data[offset : offset+8])
+	offset += 8
+	return sroar.FromBuffer(sn.data[offset : offset+deletionsLength])
```

Comment:
maybe `ByteOperations` introduced some time ago can be used here? It helps read from and write to underlying []byte and takes care for moving "cursor"

Reply:
Oh, great idea. I completely forgot that that exists. Since this is all tested well, it should be pretty simple and safe to refactor it.